spring-integration-file/src/main/java/org/springframework/integration/file/RecursiveDirectoryScanner.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.file;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.springframework.integration.file.filters.AbstractFileListFilter;
+import org.springframework.integration.file.filters.FileListFilter;
+
+/**
+ * The {@link DefaultDirectoryScanner} extension which walks through the directory tree
+ * using {@link Files#walk}.
+ *
+ * @author Artem Bilan
+ *
+ * @since 5.0
+ *
+ * @see Files#walk
+ */
+public class RecursiveDirectoryScanner extends DefaultDirectoryScanner {
+
+	@Override
+	public List<File> listFiles(File directory) throws IllegalArgumentException {
+		FileListFilter<File> filter = getFilter();
+		boolean supportAcceptFilter = filter instanceof AbstractFileListFilter;
+		try {
+			Stream<File> fileStream = Files.walk(directory.toPath())
+					.skip(1)
+					.map(Path::toFile)
+					.filter(file -> !supportAcceptFilter
+							|| ((AbstractFileListFilter<File>) filter).accept(file));
+
```

Comment:
True, but it seems a bit weird filtering a directory, but not the files beneath it.

Consider this change to your test...

```
-               this.subSubLevelFile = new File(this.subSubFolder, "file2");
+               this.subSubLevelFile = new File(this.subSubFolder, "file3");
                this.subSubLevelFile.createNewFile();
        }
 
        @Test
-       public void shouldReturnAllFilesIncludingDirs() {
-               List<File> files = new RecursiveDirectoryScanner()
-                               .listFiles(this.recursivePath.getRoot());
+       public void shouldReturnAllFilesIncludingDirs() throws IOException {
+               RecursiveDirectoryScanner scanner = new RecursiveDirectoryScanner();
+               scanner.setFilter(new AcceptOnceFileListFilter<>());
+               List<File> files = scanner.listFiles(this.recursivePath.getRoot());
                assertEquals(5, files.size());
                assertThat(files, hasItem(this.topLevelFile));
                assertThat(files, hasItem(this.subLevelFile));
                assertThat(files, hasItem(this.subSubLevelFile));
                assertThat(files, hasItem(this.subFolder));
                assertThat(files, hasItem(this.subSubFolder));
+               File file = new File(this.subSubFolder, "file4");
+               file.createNewFile();
+               files = scanner.listFiles(this.recursivePath.getRoot());
+               assertEquals(1, files.size());
+               assertThat(files, hasItem(file));
```

It just seems a bit strange that we get the directories on the first pass but not the second.